qiskit/ml/algorithms/quantum_kernel.py

```diff
@@ -0,0 +1,272 @@
+        if isinstance(quantum_instance, (BaseBackend, Backend)):
+            self._quantum_instance = QuantumInstance(quantum_instance)
+        else:
+            self._quantum_instance = quantum_instance
+
+    def _construct_circuit(self, x, y, is_statevector_sim=False):
+        """
+        Helper function to generate inner product circuit for the given feature map.
+        """
+        q = QuantumRegister(self._feature_map.num_qubits, 'q')
+        c = ClassicalRegister(self._feature_map.num_qubits, 'c')
+        qc = QuantumCircuit(q, c)
+
+        x_dict = dict(zip(self._feature_map.parameters, x))
+        psi_x = self._feature_map.assign_parameters(x_dict)
+        qc.append(psi_x.to_instruction(), qc.qubits)
+
+        if not is_statevector_sim:
+            y_dict = dict(zip(self._feature_map.parameters, y))
+            psi_y_dag = self._feature_map.assign_parameters(y_dict)
+            qc.append(psi_y_dag.to_instruction().inverse(), qc.qubits)
+
+            qc.barrier(q)
+            qc.measure(q, c)
+        return qc
+
+    def _compute_overlap(self, idx, results, is_statevector_sim):
+        """
+        Helper function to compute overlap for given input.
+        """
+        if is_statevector_sim:
+            i, j = idx
+            v_a = results.get_statevector(int(i))
+            v_b = results.get_statevector(int(j))
+            # |<0|Psi^daggar(y) x Psi(x)|0>|^2, take the amplitude
+            tmp = np.vdot(v_a, v_b)
+            kernel_value = np.vdot(tmp, tmp).real  # pylint: disable=no-member
+        else:
+            result = results.get_counts(idx)
+
+            measurement_basis = '0' * self._feature_map.num_qubits
+            kernel_value = result.get(measurement_basis, 0) / sum(result.values())
+        return kernel_value
+
+    def evaluate(self, x_vec, y_vec=None):
+        """
+        Construct kernel matrix for given data and feature map
+
+        If y is None, self inner product is calculated.
+        If using `statevector_simulator`, only build circuits for  Phi(x)|0>, then perform inner
+        product classically.
+
+        Args:
+            x_vec (numpy.ndarray): 2D array of datapoints, NxD, where N is the number of datapoints,
+                               D is the feature dimension
+            y_vec (numpy.ndarray): 2D array of datapoints, MxD, where M is the number of datapoints,
+                               D is the feature dimension
+
+        Returns:
+            numpy.ndarray: 2-D matrix, NxM
+
+        Raises:
+            ValueError:
+                - x_vec and/or y_vec are not two dimensional arrays
+                - x_vec and/or y_vec have incompatible dimension with feature map
+        """
+        if x_vec.ndim != 2:
+            raise ValueError("x_vec must be a 2D array")
+
+        if y_vec is not None and y_vec.ndim != 2:
+            raise ValueError("y_vec must be a 2D array")
+
+        if x_vec.shape[1] != self._feature_map.num_qubits:
```

Comment:
should compare `num_parameters` of feature map, not `num_qubits`

Reply:
Fixed in 64f7b26